mitodl/open-discussions
# upgrade react-redux, redux-query, react-router

#### Pre-Flight checklist


#### What are the relevant tickets?

none

#### What's this PR do?

this upgrades react-redux, redux-query, and react-router. doing so requires some changes!

the main thing is that react-redux and redux-query are no longer compatible with enzyme shallow rendering, because they use react's hooks APIs internally and hooks aren't supported in React's shallow renderer. The React team has basically said that they do not plan to add shallow rendering support for hooks, so it is up to enzyme or similar to do that if they want to. Enzyme has implemented partial support themselves, but it doesn't seem like full support is very close.

So anyway, I wanted to see how difficult it would be to move away from shallow rendering so we could clear the way for this upgrade - turns out, not so bad really. The main difficulty is that a lot of tests needed to be changed a little bit. I was able to drop `mount` in for `shallow` in the hoc renderer configuration method, wrap it in a `Router` and go from there. Surprisingly, the test suite isn't any slower! I thought it might be but was pleasantly surprised. The one thing I had to add to get things working was a method on `IntegrationTestHelper` for mocking out components - we had some tests which passed in shallow rendering because the child components weren't rendering, and whether those components render or not wasn't important to those tests anyhow, so we might as well mock them out instead of dealing with figuring out all the props they need.

Anyhow, this means we can use hooks anywhere we want now, and, in particular, we can now use the new, hooks-based APIs for redux and redux-query, which are super nice :tada: 

#### How should this be manually tested?

make sure that the app is working everywhere! and read through the code changes and make sure they make sense.
